lib/using_restricted_array.rb

```diff
@@ -1,61 +1,114 @@
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: linear or O(n), where n is the number of items in the array or the length of the input
+# Space complexity: O(1) ?
 def length(array)
-  raise NotImplementedError
+  i = 0
+  until array[i] == nil
+    i += 1
+  end
+  return i
 end
 
 # Prints each integer values in the array
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: linear or O(n), where n is the number of items in the array or the length of the input
+# Space complexity: O(1) ?
 def print_array(array)
-  raise NotImplementedError
+  len = length(array)
+  i = 0
+  while i < len
+    print array[i]
+    print ""
+    i += 1
+  end
 end
 
 # For an unsorted array, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n), where n is the number of values in the array
+# Space complexity: O(1) ?
 def search(array, length, value_to_find)
-  raise NotImplementedError
+  length.times do |i|
+    return true if array[i] == value_to_find
+  end
+
+  return false
 end
 
 # Finds and returns the largest integer value the array
 # Assumes that the array is not sorted.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n), where n is the number of values in the array
+# Space complexity: O(1) ?
 def find_largest(array, length)
-  raise NotImplementedError
+  max_value = 0
+  length.times do |i|
+    max_value = array[i] if array[i] > max_value
+  end
+
+  return max_value
 end
 
 # Finds and returns the smallest integer value in the array
 # Assumes that the array is not sorted.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n), where n is the number of values in the array
+# Space complexity: O(1) ?
 def find_smallest(array, length)
-  raise NotImplementedError
+  min_value = 222
+  length.times do |i|
+    min_value = array[i] if array[i] < min_value
+  end
+
+  return min_value
 end
 
 # Reverses the values in the integer array in place
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n), where n is the number of values in the array (not sure if this is a thing)
+# Space complexity: O(1) ?
 def reverse(array, length)
-  raise NotImplementedError
+  (length / 2).times do |i|
+    first = array[i]
+    last = array[length - (i + 1)]
+
+    array[i] = last
+    array[length - (i + 1)] = first
+  end
 end
 
 # For an array sorted in ascending order, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(log n)
+# Space complexity: O(1) ?
 def binary_search(array, length, value_to_find)
-  raise NotImplementedError
+  lowerbound = 0 # Algorithm says 1... why?
+  upperbound = length
+
+  loop do
+    if upperbound < lowerbound
+      return false
+    end
+
+    midpoint = lowerbound + (upperbound - lowerbound) / 2
+
+    # This unless thing solved things and I'm not totally sure why
```

Comment:
The `unless` block is checking for the existence of the midpoint. When your fn is searching for a non-existent value, eventually the midpoint will either be the index before the first element (`-1`) or the index after the last element. For example, say the input array is `[-1,2,3,4,7]` and you want to search for `8`. On the third iteration of the loop, `lowerbound = 5`, `upperbound=5`, and `midpoint=5`. So, `array[midpoint]=array[5]=undefined`. The unless block stops the loop by returning false before the next line evaluates. If you get rid of the unless block, the conditional `array[midpoint] < value_to_find` will break because you cannot compare `undefined` to value_to_find. Hope that helps. Another way to solve this w/o the `unless` block would be to set the upperbound to `length-1`. That will take care of `array[midpoint] < value_to_find` conditional breaking due to an undefined value.